Library/Homebrew/attestation.rb

```diff
@@ -0,0 +1,89 @@
+# typed: true
+# frozen_string_literal: true
+
+require "date"
+require "json"
+require "utils/popen"
+require "exceptions"
+
+module Homebrew
+  module Attestation
+    # @api private
+    HOMEBREW_CORE_REPO = "Homebrew/homebrew-core"
+    # @api private
+    HOMEBREW_CORE_CI_URI = "https://github.com/Homebrew/homebrew-core/.github/workflows/publish-commit-bottles.yml@refs/heads/master"
+
+    # @api private
+    BACKFILL_REPO = "trailofbits/homebrew-brew-verify"
+    # @api private
+    BACKFILL_REPO_CI_URI = "https://github.com/trailofbits/homebrew-brew-verify/.github/workflows/backfill_signatures.yml@refs/heads/main"
+
+    # No backfill attestations after this date are considered valid.
+    # @api private
+    BACKFILL_CUTOFF = DateTime.new(2024, 3, 14).freeze
+
+    # Verifies the given bottle against a cryptographic attestation of build provenance.
+    #
+    # The provenance is verified as originating from `signing_repo`, which is a `String`
+    # that should be formatted as a GitHub `owner/repo`.
+    #
+    # Callers may additionally pass in `signing_workflow`, which will scope the attestation
+    # down to an exact GitHub Actions workflow, in
+    # `https://github/OWNER/REPO/.github/workflows/WORKFLOW.yml@REF` format.
+    #
+    # @return [Hash] the JSON-decoded response.
+    # @raise [InvalidAttestationError] on any verification failures
+    #
+    # @api private
+    def self.check_attestation(bottle, signing_repo, signing_workflow = nil)
+      cmd = [HOMEBREW_GH, "attestation", "verify", bottle.cached_download, "--repo", signing_repo, "--format", "json"]
+
+      cmd += ["--cert-identity", signing_workflow] unless signing_workflow.nil?
+
+      begin
+        output = Utils.safe_popen_read(*cmd)
+      rescue ErrorDuringExecution => e
+        raise InvalidAttestationError, "attestation verification failed: #{e}"
+      end
+
+      begin
+        data = JSON.parse(output)
+      rescue JSON::ParserError
+        raise InvalidAttestationError, "attestation verification returned malformed JSON"
```

Comment:
Worth noting e.g. on which line or anything this was?

Reply:
Possibly, although at this point the JSON returned is mostly meant to be an opaque blob. We don't exactly expect this error to happen unless the user somehow replaces `gh` with something nonsensical, so perhaps it makes sense to get rid of this `InvalidAttestationError` and just let the (unlikely) `ParserError` percolate up?

Reply:
Ok, that's fine, ignore me then!